replica/table.cc

```diff
@@ -565,34 +565,37 @@ class single_storage_group_manager final : public storage_group_manager {
 
 class tablet_storage_group_manager final : public storage_group_manager {
     replica::table& _t;
+    locator::host_id _my_host_id;
+    const locator::tablet_map* _tablet_map;
 private:
-    const locator::effective_replication_map_ptr& erm() const {
-        return _t.get_effective_replication_map();
-    }
-
     const schema_ptr& schema() const {
         return _t.schema();
     }
 
-    const locator::tablet_map& tablet_map() const {
-        // FIXME: cheaper way to retrieve tablet_map than looking up every time in tablet_metadata's map.
-        auto& tm = erm()->get_token_metadata();
-        return tm.tablets().get_tablet_map(schema()->id());
+    const locator::tablet_map& tablet_map() const noexcept {
+        return *_tablet_map;
     }
 public:
-    tablet_storage_group_manager(replica::table& t) : _t(t) {}
+    tablet_storage_group_manager(table& t, const locator::effective_replication_map_ptr& erm)
+        : _t(t)
+        , _my_host_id(erm->get_token_metadata().get_my_id())
+        , _tablet_map(&erm->get_token_metadata().tablets().get_tablet_map(schema()->id()))
+    {}
```

Comment:
Since you aren't storing erm, better to pass a regular reference.

Passing a pointer: the callee may use erm for a while, blocking changes
Passing a reference: the callee must use this atomically

This is similar to how a pointer can be null and a reference cannot.

Perhaps we should rename effective_replication_map_ptr to locked_effective_replication_map_ptr to make the consequences of holding it clearer.

Reply:
I think our conventions of _ptr naming are quite clear as they are, where holding a _ptr has a "capturing" effect.
There are exceptions to this rule like `shared_sstable` that maybe should have been called `sstable_ptr`, on the lines of `schema_ptr`, `token_metadata_ptr`, and `effective_replication_map_ptr`.

Reply:
I agree about the passing a reference in this case. will change

Reply:
I routinely have to remind people to keep the ptr in scope, and to drop it when unneeded. It's clear, but it's clearly sending the wrong message. Holding on to a smart pointer keeps the value alive and doesn't impl any system-level effects. This one blocks the entire cluster from making forward progress on topology, and it's passed everywhere without a thought.

Reply:
If there was a way to make identifiers in **bold**, this would be a good use case. People would think twice before passing a **`effective_replication_map_lock`**.